### ISSUE TITLE
测试后如何可视化输出结果

### ISSUE THREAD

**Issue by zsy7532** [zhiqic/KeyPosS#1]
我运行了官方给的测试命令 CUDA_VISIBLE_DEVICES=0,1,2,3 sh tools/dist_test.sh \
    configs/face/2d_kpt_sview_rgb_img/topdown_heatmap/wflw/hrnetv2_w18_wflw_256x256_dark.py \
    exp/exp_v1.3.0/best_NME_epoch_60.pth \
    4 
并成功得到了result_keypoints.json，里面是各个关键点的坐标，但是运行后并没有可视化的图，请问我应该改动哪些地方才能得到像论文里展示的一样的可视化结果

**Comment by richardboo** [comment]
论文中关键点的可视化需要修改mmpose/apis/test.py和对应的config文件，具体可以参考如下代码。

**1. mmpose/apis/test.py:**

def single_gpu_test(model, data_loader):
    """Test model with a single gpu.

    This method tests model with a single gpu and displays test progress bar.

    Args:
        model (nn.Module): Model to be tested.
        data_loader (nn.Dataloader): Pytorch data loader.


    Returns:
        list: The prediction results.
    """

    model.eval()
    results = []
    dataset = data_loader.dataset
    prog_bar = mmcv.ProgressBar(len(dataset))
    for data in data_loader:
        with torch.no_grad():
            result = model(return_loss=False, return_heatmap=True, **data)
        results.append(result)

        # use the first key as main key to calculate the batch size
        batch_size = len(next(iter(data.values())))
        for _ in range(batch_size):
            prog_bar.update()
    
    #--------------------------------------
    # # #-------------2023-04-23------------
    if 0:
        import pdb
        # pdb.set_trace()
        mean=[0.485, 0.456, 0.406]
        std=[0.229, 0.224, 0.225]
        cnt=0
        cnt2=0
        ave = np.zeros((1,5), dtype=np.float64)
        count=0.
        for idx0, data in enumerate(data_loader):
            result = results[idx0]
            for idx1, _ in enumerate(data):
                imgs = data['img']
                targets = data['target']
                joints_3ds = data['joints_3d']
                centers = data['center']
                scales = data['scale']
                
                for idx2 in range(imgs.shape[0]):
                    # if cnt2>200:
                    #     continue
                    img = imgs[idx2]
                    joints_3d = joints_3ds[idx2][:,0:2]
                    target = targets[idx2].numpy()
                    mean_t = torch.as_tensor(mean, dtype=torch.float32, device=img.device)
                    std_t = torch.as_tensor(std, dtype=torch.float32, device=img.device)
                    img = img.permute(1,2,0)
                    img = img*std_t + mean_t

                    # input_tensor = input_tensor.squeeze()
                    # 从[0,1]转化为[0,255]，再从CHW转为HWC，最后转为cv2
                    img = img.mul_(255).add_(0.5).clamp_(0, 255).type(torch.uint8).numpy()
                    # RGB转BRG
                    img = cv2.cvtColor(img, cv2.COLOR_RGB2BGR)
                    
                    kpts = result['preds'][idx2]
                    img_path = result['image_paths'][idx2]
                    box = result['boxes'][idx2]
                    hm = result['output_heatmap'][idx2]
                    cc = centers[idx2].numpy() 
                    ss = scales[idx2].numpy()  
                    
                    sigma=2
                    pred=[]
                    # pred = keypoints_from_heatmaps_tril_img(img, np.array([target]), pred, np.array([cc]), np.array([ss]), sigma, phase=True)
                    pred, img = keypoints_from_heatmaps_tril_img(img, np.array([hm]), pred, np.array([cc]), np.array([ss]), sigma, phase=True)
                    pred=pred[0]

                    # img = cv2.imread(img_path)

                    for index, pt in enumerate(kpts):
                        if index != 16:
                            continue
                        # x=int(pt[0])
                        # y=int(pt[1])
                        x=int(pred[index][0])
                        y=int(pred[index][1])
                        cv2.circle(img, (x, y), 3, (0,0,255), -1)
                     
                        x2=int(joints_3d[index][0])
                        y2=int(joints_3d[index][1])
                        cv2.circle(img, (x2, y2), 3, (0,255,0), -1)

                        cv2.line(img, (x,y), (x2, y2), (255,0,0) )

                        # x1=int(box[0])
                        # y1=int(box[1])
                        # x2=int(box[2])
                        # y2=int(box[3])
                        # img = cv2.rectangle(img, (x1, y1), (x2, y2), (0,255,0), 2)
                    cnt +=1
                    if cnt%20==0 and cnt2<100:
                        img_path = f'./exp/exp888/show/test_{cnt}.jpg'
                        cv2.imwrite(img_path, img)
                        cnt2 +=1
    
    #--------------------------------------

    return results

2. 修改config文件，以configs\face\2d_kpt_sview_rgb_img\topdown_heatmap\cofw\hrnetv2_w18_cofw_256x256.py为例，修改val_pipeline字段
val_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='TopDownGetBboxCenterScale', padding=1.25),
    dict(type='TopDownAffine'),
    dict(type='ToTensor'),
    dict(
        type='NormalizeTensor',
        mean=[0.485, 0.456, 0.406],
        std=[0.229, 0.224, 0.225]),
    dict(
        type='Collect',
        keys=['img', 'joints_3d', 'center', 'scale'],
        meta_keys=['image_file', 'center', 'scale', 'rotation', 'flip_pairs']),
]
